### PR TITLE
Benchmark

### PR DESCRIPTION
**Pull request by djc**

### CODE REVIEW

**Comment by Ralith** [comment]
Runs now (though it panics immediately).

**Comment by djc** [comment]
Rebased onto current nightly. Off the top of my head, not yet sure how to deal with the `Runtime` borrowing issues.

File: quinn/benches/bench.rs [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+    runtime.spawn(server_driver.map_err(|e| panic!("server driver failed: {}", e)));
+    runtime.spawn(client_driver.map_err(|e| panic!("client driver failed: {}", e)));
+
+    let runtime = Rc::new(RefCell::new(runtime));
+    {
+        let nested = runtime.clone();
+        runtime
+            .borrow_mut()
+            .spawn(server_incoming.for_each(move |(driver, _, incoming)| {
+                nested.borrow_mut().spawn(driver.map_err(|_| panic!()));
```
Comment (Ralith) [inline quinn/benches/bench.rs]: Replacing this with `tokio::runtime::current_thread::spawn` should do the trick, I think.

**Comment by djc** [comment]
@Ralith would you mind taking a look? This relatively straightforward translation to futures-0.1 seems to mostly work except that cleanup doesn't work correctly yet and I haven't been able to figure out why exactly.

**Comment by Ralith** [comment]
Will investigate.

**Comment by Ralith** [comment]
The benchmark was failing due to genuine connection ID backpressure, caused by the `ReadAllUnordered` future incorrectly returning `NotReady`. A future that is built out of other poll operations must never return `NotReady` unless one or more of its components returned `NotReady`, because a poll operation that returns `Ready` does not guarantee that the task will be woken again in the future.

**Comment by djc** [comment]
I propose to merge this with the single benchmark for now, which is a good start and can inform further optimizations (and pulling it into other branches to measure impace is a pain). Left the second one commented out so we can enable it separately.

**Pull request merged**